addons/stock/models/stock_move.py

```diff
@@ -1724,3 +1752,63 @@ def _adjust_procure_method(self):
     def _show_details_in_draft(self):
         self.ensure_one()
         return self.state != 'draft' or (self.picking_id.immediate_transfer and self.state == 'draft')
+
+    @api.depends('move_line_ids', 'move_line_ids.lot_id', 'move_line_ids.qty_done')
+    def _lot_ids_compute(self):
+        for move in self:
+            if move.picking_type_id.show_reserved is False:
+                move.lot_ids = move.move_line_nosuggest_ids.filtered(lambda ml: ml.lot_id and not float_is_zero(ml.qty_done, precision_rounding=ml.product_uom_id.rounding)).lot_id
+            else:
+                move.lot_ids = move.move_line_ids.filtered(lambda ml: ml.lot_id and not float_is_zero(ml.qty_done, precision_rounding=ml.product_uom_id.rounding)).lot_id
+
+    def _lot_ids_inverse(self):
+        for move in self:
+            move_lines_commands = []
+            if move.picking_type_id.show_reserved is False:
+                mls = move.move_line_nosuggest_ids
+            else:
+                mls = move.move_line_ids
+            mls = mls.filtered(lambda ml: ml.lot_id)
+            for ml in mls:
+                if not ml.lot_id in move.lot_ids:
+                    move_lines_commands.append((2, ml.id))
+            ls = move.move_line_ids.lot_id
+            for lot in move.lot_ids:
+                if not lot in ls:
+                    move_line_vals = self._prepare_move_line_vals(quantity=0)
+                    move_line_vals['lot_id'] = lot.id
+                    move_line_vals['lot_name'] = lot.name
+                    move_line_vals['qty_done'] = 1
+                    move_lines_commands.append((0, 0, move_line_vals))
+            move.write({'move_line_ids': move_lines_commands})
+
+    @api.onchange('lot_ids')
+    def _onchange_lot_ids(self):
+        already_used_lots = []
+        for lot in self.lot_ids:
+            counter = self.env['stock.move.line'].search_count([
+                ('company_id', '=', self.company_id.id),
+                ('product_id', '=', self.product_id.id),
+                ('lot_id', '=', lot._origin.id),
+                ('move_id', 'in', self.env['stock.move'].search([
+                    ('company_id', '=', self.company_id.id),
+                    ('product_id', '=', self.product_id.id),
+                    ('state', '!=', 'cancel'),
+                    ('id', '!=', self._origin.id)
+                    ])._ids
+                ),   
+            ])
+            if counter > 0:
+                already_used_lots.append(lot.name)
+        counter = self.env['stock.move.line'].search_count([
```

Comment:
counter  = self.move_line_ids.filtered(lambda line: line.lot_name and not line.lot_id)
it search in the cache instead of the database